jobsdb/jobsdb_test.go

```diff
@@ -303,3 +312,94 @@ var dsListInMemory = []dataSetT{
 	d1,
 	d2,
 }
+
+func BenchmarkSanitizeJson(b *testing.B) {
+	size := 4_000_000
+	nulls := 100
+
+	// string with nulls
+	inputWithoutNulls := randomString(size - nulls*len(`\u0000`))
+	inputWithNulls := insertStringInString(inputWithoutNulls, `\u0000`, nulls)
+	require.Equal(b, inputWithoutNulls, sanitizedJsonUsingStrings(json.RawMessage(inputWithNulls)))
+	require.Equal(b, inputWithoutNulls, sanitizedJsonUsingBytes(json.RawMessage(inputWithNulls)))
+	require.Equal(b, inputWithoutNulls, sanitizedJsonUsingRegexp(json.RawMessage(inputWithNulls)))
+	b.Run(fmt.Sprintf("SanitizeUsingStrings string of size %d with null characters", size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			sanitizedJsonUsingStrings(json.RawMessage(inputWithNulls))
+		}
+	})
+	b.Run(fmt.Sprintf("SanitizeUsingBytes string of size %d with null characters", size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			sanitizedJsonUsingBytes(json.RawMessage(inputWithNulls))
+		}
+	})
+	b.Run(fmt.Sprintf("SanitizeUsingRegexp string of size %d with null characters", size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			sanitizedJsonUsingRegexp(json.RawMessage(inputWithNulls))
+		}
+	})
+
+	// string without null characters
+	input := randomString(size)
+	b.Run(fmt.Sprintf("SanitizeUsingStrings string of size %d without null characters", size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			sanitizedJsonUsingStrings(json.RawMessage(input))
+		}
+	})
+	b.Run(fmt.Sprintf("SanitizeUsingBytes string of size %d without null characters", size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			sanitizedJsonUsingBytes(json.RawMessage(input))
+		}
+	})
+	b.Run(fmt.Sprintf("SanitizeUsingRegexp string of size %d without null characters", size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			sanitizedJsonUsingRegexp(json.RawMessage(input))
+		}
+	})
+
+}
+
+func randomString(n int) string {
+	var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
+	s := make([]rune, n)
+	for i := range s {
+		s[i] = letters[rand.Intn(len(letters))]
+	}
+	return string(s)
+}
+
+func insertStringInString(input string, c string, times int) string {
+	if times == 0 {
+		return input
+	}
+	pos := map[int]struct{}{}
+	for len(pos) < times {
+		newPos := rand.Intn(len(input))
+		pos[newPos] = struct{}{}
+	}
+	keys := make([]int, 0, len(pos))
+	for k := range pos {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
+	offset := len(c)
+	for i, idx := range keys {
+		oidx := idx + i*offset
+		input = input[:oidx] + c + input[oidx:]
+	}
+	return input
+}
+
+func sanitizedJsonUsingStrings(input json.RawMessage) string {
+	return strings.Replace(string(input), `\u0000`, "", -1)
+}
+
+func sanitizedJsonUsingBytes(input json.RawMessage) string {
```

Comment:
I believe we are being "unfair" to bytes benchmark here.

In your tests, you convert it back to a string. However in the actual implementation, you return `json.RawMessage`, which is a []byte.

Not very important, I just expected bytes.ReplaceAll to be better and then check the benchs :)

Reply:
Yes you are correct, but even after being fair, `strings` are still winning when there are no null characters

```
BenchmarkSanitizeJson/SanitizeUsingStrings_string_of_size_4000000_with_null_characters-10         	    1186	    972290 ns/op	12017684 B/op	       3 allocs/op
BenchmarkSanitizeJson/SanitizeUsingBytes_string_of_size_4000000_with_null_characters-10           	    1834	    643793 ns/op	 8011793 B/op	       2 allocs/op
BenchmarkSanitizeJson/SanitizeUsingRegexp_string_of_size_4000000_with_null_characters-10          	     544	   2163382 ns/op	31734351 B/op	      51 allocs/op
BenchmarkSanitizeJson/SanitizeUsingStrings_string_of_size_4000000_without_null_characters-10      	    2082	    571437 ns/op	 8011797 B/op	       2 allocs/op
BenchmarkSanitizeJson/SanitizeUsingBytes_string_of_size_4000000_without_null_characters-10        	    1610	    731277 ns/op	 8011789 B/op	       2 allocs/op
BenchmarkSanitizeJson/SanitizeUsingRegexp_string_of_size_4000000_without_null_characters-10       	    1012	   1190263 ns/op	16029709 B/op	      12 allocs/op
```